compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala

```diff
@@ -0,0 +1,358 @@
+        else
+          val indices = annotated
+            .paramSymss
+            .zipWithIndex
+            .flatMap: (paramClause, paramClauseIndex) =>
+              val annotationIndices = findUnrollAnnotations(paramClause)
+              if (annotationIndices.isEmpty) None
+              else Some((paramClauseIndex, annotationIndices))
+          if indices.nonEmpty then
+            // pre-validation should have occurred in posttyper
+            assert(annotated.is(Final, butNot = Deferred) || annotated.isConstructor || annotated.owner.is(ModuleClass) || annotated.name.is(DefaultGetterName),
+              i"$annotated is not final&concrete, or a constructor")
+          indices
+      })
+    end computeIndices
+
+    override def transform(tree: tpd.Tree)(using Context): tpd.Tree = tree match
+      case tree @ TypeDef(_, impl: Template) if classes(tree.symbol) =>
+        super.transform(cpy.TypeDef(tree)(rhs = unrollTemplate(impl, computeIndices)))
+      case tree =>
+        super.transform(tree)
+  }
+
+  def copyParamSym(sym: Symbol, parent: Symbol)(using Context): (Symbol, Symbol) =
+    val copied = sym.copy(owner = parent, flags = (sym.flags &~ HasDefault), coord = sym.coord)
+    sym -> copied
+
+  def symLocation(sym: Symbol)(using Context) = {
+    val lineDesc =
+      if (sym.span.exists && sym.span != sym.owner.span)
+        s" at line ${sym.srcPos.line + 1}"
+      else ""
+    i"in ${sym.owner}${lineDesc}"
+  }
+
+  def findUnrollAnnotations(params: List[Symbol])(using Context): List[Int] = {
+    params
+      .zipWithIndex
+      .collect {
+        case (v, i) if v.hasAnnotation(defn.UnrollAnnot) =>
+          i
+      }
+  }
+
+  def isTypeClause(p: ParamClause) = p.headOption.exists(_.isInstanceOf[TypeDef])
+
+  def generateSingleForwarder(defdef: DefDef,
+                              prevMethodType: Type,
+                              paramIndex: Int,
+                              paramCount: Int,
+                              nextParamIndex: Int,
+                              nextSymbol: Symbol,
+                              annotatedParamListIndex: Int,
+                              isCaseApply: Boolean)(using Context) = {
+
+    def initNewForwarder()(using Context): (TermSymbol, List[List[Symbol]]) = {
+      val forwarderDefSymbol0 = Symbols.newSymbol(
+        defdef.symbol.owner,
+        defdef.name,
+        defdef.symbol.flags &~ HasDefaultParams |
+        Invisible | Synthetic,
+        NoType, // fill in later
+        coord = nextSymbol.span.shift(1) // shift by 1 to avoid "secondary constructor must call preceding" error
+      ).entered
+
+      val newParamSymMappings = extractParamSymss(copyParamSym(_, forwarderDefSymbol0))
+      val (oldParams, newParams) = newParamSymMappings.flatten.unzip
+
+      val newParamSymLists0 =
+        newParamSymMappings.map: pairss =>
+          pairss.map: (oldSym, newSym) =>
```

Comment:
`pairss` doesn't appear to be a list of lists, is it? If not, call it `pairs`:
```suggestion
        newParamSymMappings.map: pairs =>
          pairs.map: (oldSym, newSym) =>
```